controller/object_store_controller.go

```diff
@@ -0,0 +1,1139 @@
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+	} else if *dpl.Spec.Replicas != 0 {
+		dpl.Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// Wait for all directly owmed resources to be deleted before removing the finalizer
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetIngress(osc.namespace, store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetDeployment(store.Name)
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err == nil || !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	return osc.ds.RemoveFinalizerForObjectStore(store)
+}
+
+func (osc *ObjectStoreController) initializeObjectStore(store *longhorn.ObjectStore) (err error) {
+	if !(store.Spec.TargetState == longhorn.ObjectStoreStateStopped) {
+		store.Status.State = longhorn.ObjectStoreStateStarting
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreatePVC(store *longhorn.ObjectStore) (*corev1.PersistentVolumeClaim, *longhorn.ObjectStore, error) {
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err == nil {
+		return pvc, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pvc, err = osc.createPVC(store)
+		if err != nil {
+			store.Status.State = longhorn.ObjectStoreStateError
+			return nil, store, errors.Wrap(err, "failed to create persistent volume claim")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
```

Comment:
Any explanations for this?

Reply:
The link points to nothing, so it's not clear to me what the suggestion is. I assume that the suggestion is to set the state in `handleStarting`?
In that case I don't see much benefit in moving that. This status transition is for the unlikely case that an object store ends up in state `error`, due to that, the `handleStarting` function is executed and a resource is recreated. In that case, I think it's wrong to remain in `error` state, since the controller tried to fix something, but we have to keep executing the `handleStarting` function in case other resources are missing also and need to be recreated and so the controller waits for all resources to become healthy before transitioning to `running` state.